vortex-datafusion/src/persistent/execution.rs

```diff
@@ -151,20 +151,36 @@ impl ExecutionPlan for VortexExec {
     }
 }
 
-fn repartition_by_count(
+fn repartition_by_size(
     file_groups: Vec<Vec<PartitionedFile>>,
     desired_partitions: usize,
 ) -> Vec<Vec<PartitionedFile>> {
     let all_files = file_groups.into_iter().concat();
+    let total_size = all_files.iter().map(|f| f.object_meta.size).sum::<usize>();
+    let target_partition_size = total_size / (desired_partitions + 1);
 
-    let approx_files_per_partition = all_files.len().div_ceil(desired_partitions);
-    let mut repartitioned_file_groups = Vec::default();
+    let mut partitions = Vec::with_capacity(desired_partitions);
 
-    for chunk in &all_files.into_iter().chunks(approx_files_per_partition) {
-        repartitioned_file_groups.push(chunk.collect::<Vec<_>>());
+    let mut curr_partition_size = 0;
+    let mut curr_partition = Vec::default();
+
+    for file in all_files.into_iter() {
+        curr_partition_size += file.object_meta.size;
+        curr_partition.push(file);
+
+        if curr_partition_size > target_partition_size {
+            curr_partition_size = 0;
+            partitions.push(std::mem::take(&mut curr_partition));
+        }
+    }
+
+    // if there's anything left, we shove it into existing partitions
+    for (idx, file) in curr_partition.into_iter().enumerate() {
+        let part_idx = idx % partitions.len();
+        partitions[part_idx].push(file);
     }
```

Comment:
can you add a check that `partitions.len() == desired_partitions` and otherwise panic? because this should/will be the case unless we introduce a math bug